libraries/Microsoft.Bot.Builder.LanguageGeneration/TemplateEngine.cs

```diff
@@ -47,27 +47,40 @@ public TemplateEngine()
         /// <returns>Teamplate engine with parsed files.</returns>
         public TemplateEngine AddFiles(IEnumerable<string> filePaths, ImportResolverDelegate importResolver = null)
         {
+            var filePathContainer = new HashSet<string>();
+
             foreach (var filePath in filePaths)
             {
                 importResolver = importResolver ?? ((id) =>
                  {
-                    // import paths are in resource files which can be executed on multiple OS environments
-                    // Call GetOsPath() to map / & \ in importPath -> OSPath
-                    var importPath = GetOsPath(id);
-                    if (!Path.IsPathRooted(importPath))
+                     // import paths are in resource files which can be executed on multiple OS environments
+                     // Call GetOsPath() to map / & \ in importPath -> OSPath
+                     var importPath = GetOsPath(id);
+                     if (!Path.IsPathRooted(importPath))
                      {
-                        // get full path for importPath relative to path which is doing the import.
-                        importPath = Path.GetFullPath(Path.Combine(Path.GetDirectoryName(filePath), id));
+                         // get full path for importPath relative to path which is doing the import.
+                         importPath = Path.GetFullPath(Path.Combine(Path.GetDirectoryName(filePath), id));
                      }
-                    return (File.ReadAllText(importPath), importPath);
+
+                     if (!filePathContainer.Contains(importPath))
```

Comment:
We should keep `importResolver ` pure, and leave no other logic code in this delegate.  The only thing this delegate do is transfer from 'literal path' to 'file path'.

Reply:
current importResolver also read file content and return file content and file path. You mean we change it to return file path only?

Reply:
Removed fileContainer from import resolver